src/modules/job-info/update.c

```diff
@@ -0,0 +1,720 @@
+     * called via zlist_remove() */
+    zhashx_delete (ctx->index_uw, uc->index_key);
+    zlist_remove (ctx->update_watchers, uc);
+}
+
+static int eventlog_watch (struct update_ctx *uc)
+{
+    const char *topic = "job-info.eventlog-watch";
+    int rpc_flags = FLUX_RPC_STREAMING;
+    flux_msg_t *msg = NULL;
+    int save_errno, rc = -1;
+
+    if (!(uc->eventlog_watch_f = flux_rpc_pack (uc->ctx->h,
+                                                topic,
+                                                FLUX_NODEID_ANY,
+                                                rpc_flags,
+                                                "{s:I s:s s:i}",
+                                                "id", uc->id,
+                                                "path", "eventlog",
+                                                "flags", 0))) {
+        flux_log_error (uc->ctx->h, "%s: flux_rpc_pack", __FUNCTION__);
+        goto error;
+    }
+
+    if (flux_future_then (uc->eventlog_watch_f,
+                          -1,
+                          eventlog_continuation,
+                          uc) < 0) {
+        /* future cleanup handled with context destruction */
+        flux_log_error (uc->ctx->h, "%s: flux_future_then", __FUNCTION__);
+        goto error;
+    }
+
+    rc = 0;
+error:
+    save_errno = errno;
+    flux_msg_destroy (msg);
+    errno = save_errno;
+    return rc;
+}
+
+static void lookup_continuation (flux_future_t *f, void *arg)
+{
+    struct update_ctx *uc = arg;
+    struct info_ctx *ctx = uc->ctx;
+    const char *key_str;
+    const char *eventlog_str;
+    const char *input;
+    const char *tok;
+    size_t toklen;
+    const char *errmsg = NULL;
+    bool job_ended = false;
+    bool submit_parsed = false;
+    const flux_msg_t *msg;
+
+    if (flux_rpc_get_unpack (f, "{s:s s:s}",
+                             uc->key, &key_str,
+                             "eventlog", &eventlog_str) < 0) {
+        if (errno != ENOENT && errno != EPERM)
+            flux_log_error (ctx->h, "%s: flux_rpc_get_unpack", __FUNCTION__);
+        goto error;
+    }
+
+    /* if count == 0, all callers canceled streams */
+    if (flux_msglist_count (uc->msglist) == 0)
+        goto cleanup;
+
+    if (!(uc->update_object = json_loads (key_str, 0, NULL))) {
+        errno = EINVAL;
+        errmsg = "lookup value cannot be parsed";
+        goto error;
+    }
+
+    input = eventlog_str;
+    while (eventlog_parse_next (&input, &tok, &toklen)) {
+        json_t *entry = NULL;
+        const char *name;
+        json_t *context = NULL;
+        if (eventlog_parse_entry_chunk (uc->ctx->h,
+                                        tok,
+                                        toklen,
+                                        &entry,
+                                        &name,
+                                        &context) < 0) {
+            errmsg = "error parsing eventlog";
+            goto error;
+        }
```

Comment:
Yeah, as noted earlier, it looks like this could be replaced with (with appropriate error handling)
```c
eventlog = eventlog_decode (eventlog_str);
json_array_foreach (eventlog, index, entry) {
    eventlog_entry_parse (entry, NULL, &name, &context);
    ...
}
```